src/utils/userLocation.ts

```diff
@@ -0,0 +1,83 @@
+import firebase, { UserInfo } from "firebase/app";
+
+import { AnyRoom } from "types/rooms";
+
+import { updateUserProfile } from "pages/Account/helpers";
+import { useInterval } from "hooks/useInterval";
+
+import { getCurrentTimeInMilliseconds } from "./time";
+import { openRoomUrl, openUrl, venueInsideUrl } from "./url";
+
+const LOCATION_INCREMENT_SECONDS = 10;
+const LOCATION_INCREMENT_MS = LOCATION_INCREMENT_SECONDS * 1000;
+
+type LocationData = Record<string, number>;
+
+export const updateLocationData = (
+  userId: string,
+  newLocationData: LocationData = {}
+) => {
+  updateUserProfile(userId, {
+    lastSeenAt: getCurrentTimeInMilliseconds(),
+    lastSeenIn: newLocationData,
+  });
+};
+
+export interface TrackLocationProps {
+  userId: string;
+  locationName: string;
+  lastSeenIn?: LocationData;
+}
+
+export const trackLocationEntered = ({
+  userId,
+  locationName,
+}: TrackLocationProps) => {
+  updateLocationData(userId, {
+    [locationName]: getCurrentTimeInMilliseconds(),
+  });
+};
+
+export const enterVenue = (venueId: string) => openUrl(venueInsideUrl(venueId));
+
+type EnterExternalRoomProps = {
+  userId: string;
+  room: AnyRoom;
+};
+
+export const enterExternalRoom = ({ userId, room }: EnterExternalRoomProps) => {
+  trackLocationEntered({
+    userId,
+    locationName: room.url,
+  });
+
+  openRoomUrl(room.url);
+};
+
+export const useUpdateTimespentPeriodically = (
+  user: UserInfo | undefined,
+  roomName: string
+) => {
+  const shouldUseInterval = user && roomName;
+
+  useInterval(
+    () => {
+      // Time spent is currently counted multiple time if multiple tabs are open
+      if (!user || !roomName) return;
+
+      const firestore = firebase.firestore();
+      const doc = `users/${user.uid}/visits/${roomName}`;
+      const increment = firebase.firestore.FieldValue.increment(
+        LOCATION_INCREMENT_SECONDS
+      );
+
+      return firestore
+        .doc(doc)
+        .update({ timeSpent: increment })
```

Comment:
Since this field is only used once, it seems like it would be cleaner if it was just included directly in the `.update()` rather than split out as it's own `const`

Reply:
Fixed small things regarding this chunk of code, but I'd call it OOS for this PR since this is not related to counting

Reply:
Since you renamed the file/etc, all I see is an entire file of changed diff, so I comment appropriately.